src/transformStateWithClones.js

```
@@ -7,7 +7,49 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties': {
+        const extraData =
+          action &&
+          typeof action.extraData === 'object' &&
+          action.extraData !== null &&
+          !Array.isArray(action.extraData)
+            ? action.extraData
+            : {};
+
+        currentState = {
+          ...currentState,
+          ...extraData,
+        };
+        break;
+      }
+
+      case 'removeProperties':
+        currentState = { ...currentState };
+
+        if (action.keysToRemove && Array.isArray(action.keysToRemove)) {
+          for (const key of action.keysToRemove) {
+            delete currentState[key];
+          }
+        }
+        break;
+
+      default:
+        continue;
```

This violates checklist item 2.4 and 3.5: 'The switch statement must include a default case for error handling (e.g., to handle unknown action types)'. Using `continue;` silently skips unknown actions, which can hide bugs. You should throw an error or handle unknown actions explicitly.